### ISSUE TITLE
Error: Expected shape from model of {} does not match actual shape of {1,1,1} for output 

### ISSUE THREAD

**Issue by realjanpaulus** [huggingface/optimum#85]
### Problem
I'm getting the following error when I'm trying to apply **static quantization** (ONNX) with the `ORTQuantizer` . 

<img width="1587" alt="image" src="https://user-images.githubusercontent.com/22560883/156150800-b426a764-f922-47ec-b0b8-b2d991094f32.png">

### Tests
This error occurs for:
- my custom script
- the example code in the [README.md](https://github.com/huggingface/optimum#quickstart)
-  The example [notebook] in this repository (https://github.com/huggingface/notebooks/blob/master/examples/text_classification_quantization_ort.ipynb)
- <img width="1072" alt="image" src="https://user-images.githubusercontent.com/22560883/156150897-0aba6c24-7044-47a4-ba52-cf3fcc60f7fb.png">
- a brand new project with only `transformers`, `datasets` and `optimum[onnxruntime]` installed
- a brand new project with only `transformers`, `datasets` and `optimum[onnxruntime]` (with `python -m pip install git+https://github.com/huggingface/optimum.git` installed)

### More

- The resulting `model-quantized.onnx` can be loaded but produces very bad results. 
- **dynamic quantization** works seamlessy
- using:
    - Python 3.9
    - tested on two different devices with different operating systems: 
        - MacOS Monterey (with Intel)
        - WSL for Windows 11 (Ubuntu)

**Comment by severinsimmler** [comment]
@realjanpaulus Can you add a code snippet with a minimal working example?

**Comment by realjanpaulus** [comment]
@severinsimmler of course, similar to the example in this [notebook](https://github.com/huggingface/notebooks/blob/master/examples/text_classification_quantization_ort.ipynb)

```python
from optimum.onnxruntime import ORTConfig, ORTQuantizer
from transformers import AutoTokenizer, DataCollatorWithPadding

approach = "static"

model_ckpt = "distilbert-base-uncased-finetuned-sst-2-english"
tokenizer = AutoTokenizer.from_pretrained(model_ckpt)
ort_config = ORTConfig(quantization_approach=approach)

def preprocess_fn(ex):
    return tokenizer(ex["sentence"])

data_collator = DataCollatorWithPadding(tokenizer)
if approach == "static":
    quantizer = ORTQuantizer(
        ort_config,
        dataset_name="glue",
        dataset_config_name="sst2",
        preprocess_function=preprocess_fn,
        data_collator=data_collator,
        per_channel=True,
    )
else:
    quantizer = ORTQuantizer(ort_config)
quantizer.fit(model_ckpt, output_dir=".", feature="sequence-classification")

```

**Comment by echarlaix** [comment]
Hi @realjanpaulus, 

This warning should not have any impact on the final performance and should be fixed shortly by the ONNX Runtime team.
For more informations, you can take a look at the following issue : https://github.com/microsoft/onnxruntime/issues/10504
> The resulting model-optimized.onnx can be loaded but produces very bad results.

Concerning the resulting model-quantized.onnx (model-optimized.onnx should be obtained when optimizing the model through the utilization of an `ORTOptimizer`) accuracy drop, you might want to take a look at the types of operators to quantize (by specifying `op_types_to_quantize` when instantiating an `ORTConfig`) as done in the [notebook](https://github.com/huggingface/notebooks/blob/master/examples/text_classification_quantization_ort.ipynb).

**Comment by realjanpaulus** [comment]
Hi @echarlaix, 
Thank you first of all for the quick answer. Sorry, I meant `model-quantized.onnx`, not `model-optimized.onnx`. I fixed this within the issue. I have another question: 

How important is the calibration dataset? In your [run_qa.py](https://github.com/huggingface/optimum/blob/9abdfc21cfa0befd7c3f50805812c677ffd3adde/examples/onnxruntime/pytorch/question-answering/run_qa.py#L702-L727) example, you used one of the dataset splits for the calibration dataset. Is the rule here, the more data, the more precise? Or does a small dataset handle the job quite well (or even something like `dataset = Dataset.from_dict({"context": ["This is a context", "This is another"]})`)

I also did some experiments where I tuned a lot of the `ORTConfig` parameters and these are the results I got, maybe they are interesting for you or other people which stumble across similar problems:

- `calibration_method = "percentile"` does work but give very bad results, even when using only something like `op_types_to_quantize=["MatMul"]`
- `calibration_method = "entropy"` is the only calibration method that works
    - regarding `op_types_to_quantize` here: the less parameter, the higher the accuracy and the model size (which makes sense)
    - below are some experiment results in a table form I did for my task I'm currently working. I can't share too much insight about the task, but I fine-tuned an `XLMRoberta Base` model for `QuestionAnswering` which I would like to convert to ONNX for a faster **inference**. The unquantized/unoptimized `model.onnx` has a size of ~1100mb
        |  | `op_types_to_quantize` | Precision | Recall | Size (MB) |
        | -------- | -------- | -------- | -------- | -------- |
        | **dynamic** | `None` | 84.23 | 91.32 | 278.4 |
        | **static** | `None` | 12.28 | 6.32 | 278.1 |
        | **static** | `["MatMul"]` | 83.73 | 92.18 | 855.6 |
        | **static** | `["Mul"]` | 85.59 | 88.95 | 855.6 |
        | **static** | `["Reshape"]` | 85.20 | 91.32 | ~1100 |
        | **static** | `["Gather"]` | 84.84 | 88.95  | 532.7 |
        | **static** | `["Transpose"]` | 85.28 | 91.32 | ~1100 |
        | **static** | `["MatMul", "Reshape"]` | 83.73 | 92.18 | 855.6 |
        | **static** | `["MatMul", "Transpose"]` | 83.73 | 92.18 | - |

**Comment by mfuntowicz** [comment]
Hi @realjanpaulus, 

We reported the issue to ORT folks and should be fixed in next release 👍🏻.

In the meantime I confirm it doesn't impact final performances of the model: 

↪️  https://github.com/microsoft/onnxruntime/issues/10504

**Comment by echarlaix** [comment]
Hi @realjanpaulus,

Thanks for sharing your experiments results, analyzing which part of the model is sensitive to quantization is a very interesting topic.
 > Is the rule here, the more data, the more precise? Or does a small dataset handle the job quite well (or even something like dataset = Dataset.from_dict({"context": ["This is a context", "This is another"]}))

I would tend to say that in general the more calibration data you provide, the more confident we can be in the estimated quantization parameters. This is however not true for calibration methods such as minmax which takes the global minimum and maximum values (there is currently not the option to compute those values using an exponential moving average). This results in an increase of the quantization range, leading to a decrease in precision and very likely a drop in the final model's performance.
Even though this should not be true for every model / task / calibration method combination, I found that for BERT models on text classification tasks, 40 to 50 examples were giving good results when using the minmax calibration method.

I hope this helps !

I will close this issue as the initial problem is now solved, if you have other questions please feel free to open an other one.
